### PR TITLE
GET: feedback/score

### LINKED ISSUE #14: add GET: feedback/score
similar to /report/score we want the score but for feedbacks

### PR DESCRIPTION
**Pull request by RusticPotatoes**
for get feedback https://github.com/Bot-detector/public-api/issues/14

### CODE REVIEW

File: mysql/docker-entrypoint-initdb.d/02_data.sql [hunk 1]
```diff
@@ -15,7 +15,9 @@ INSERT INTO
         ironman,
         hardcore_ironman,
         ultimate_ironman,
-        normalized_name
+        normalized_name,
+        feedback_received,
+        feedback_given
```
Comment (extreme4all) [inline mysql/docker-entrypoint-initdb.d/02_data.sql]: this does not exist in Players table

File: mysql/docker-entrypoint-initdb.d/02_data.sql [hunk 2]
```diff
@@ -356,4 +365,88 @@ CREATE PROCEDURE INSERTROWS(NUM INT) BEGIN
...
\ No newline at end of file
+CALL InsertRows(250);
+
+CREATE TABLE `Labels` (
+  `id` int NOT NULL AUTO_INCREMENT,
+  `label` varchar(50) CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci NOT NULL,
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `Unique_label` (`label`) USING BTREE
+)
+;
```
Comment (extreme4all) [inline mysql/docker-entrypoint-initdb.d/02_data.sql]: there is a file for tables :)

File: mysql/docker-entrypoint-initdb.d/02_data.sql [hunk 3]
```diff
@@ -356,4 +365,88 @@ CREATE PROCEDURE INSERTROWS(NUM INT) BEGIN
...
+    pr.predicted_confidence,
+    "" AS feedback_text,
+    CASE WHEN RAND() > 0.5 THEN 1 ELSE -1 END AS vote,
+    (SELECT label FROM Labels ORDER BY RAND() LIMIT 1) AS proposed_label
+FROM (SELECT * FROM Players ORDER BY RAND() LIMIT 1000) pl1
+JOIN (SELECT * FROM Players ORDER BY RAND() LIMIT 1000) pl2 ON pl1.id <> pl2.id
+JOIN Predictions pr ON pr.id = pl2.id
+ORDER BY RAND()
+LIMIT 100
+;
```
Comment (extreme4all) [inline mysql/docker-entrypoint-initdb.d/02_data.sql]: select won't do anything please remove

File: src/app/models/feedback.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+from sqlalchemy.sql.expression import select
+from sqlalchemy.orm import aliased
+from sqlalchemy.engine import Result
+from src.core.database.models.feedback import DataModelPredictionFeedback as dbFeedback
+from src.core.database.models.player import Player as dbPlayer
+from src.app.views.response.feedback import PredictionFeedbackResponse
+import logging
+
+
+class AppModelFeedback:
```
Comment (extreme4all) [inline src/app/models/feedback.py]: rename to Feedback

File: src/app/models/feedback.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    def __init__(self, session: AsyncSession) -> None:
+        self.session = session
+        self.logger = logging.getLogger("AppModelFeedback")
+
+    async def get_feedback_responses(self, player_names: list[str]):
+        async with self.session:
+            print(f"Player names: {player_names}")
+            fv: dbFeedback = aliased(dbFeedback, name="feedback_voter")
+            fs: dbFeedback = aliased(dbFeedback, name="feedback_subject")
+            pn: dbPlayer = aliased(dbPlayer, name="player_name")
```
Comment (extreme4all) [inline src/app/models/feedback.py]: this could be a bit more verbose, fb_voter, fb_subject, pl_name

File: src/app/views/input/feedback.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    vote: int = Field(..., ge=-1, le=1, description="Vote is -1, 0 or 1")
+    prediction: str = Field(..., example="Real_Player")
+    confidence: Optional[float] = Field(0, ge=0, le=1, description="double from 0 to 1")
+    subject_id: int = Field(..., example=1)
+    feedback_text: Optional[str] = Field(None, example="Test feedback")
+    proposed_label: Optional[str] = Field(None, example="Real_Player")
+
+
+class PredictionFeedbackOut(BaseModel):
+    message: str = Field(..., example="Feedback submitted successfully")
```
Comment (extreme4all) [inline src/app/views/input/feedback.py]: not needed

File: mysql/docker-entrypoint-initdb.d/02_data.sql (hunk 3 above)
Comment (RusticPotatoes) [reply mysql/docker-entrypoint-initdb.d/02_data.sql]: yep you are right

File: src/app/models/feedback.py (hunk 4 above)
Comment (RusticPotatoes) [reply src/app/models/feedback.py]: will do, was hardcore debugging, and feedback being all the class names was not helping

Comment (RusticPotatoes) [reply src/app/models/feedback.py]: resolved by [b9f3d53](https://github.com/Bot-detector/public-api/pull/25/commits/b9f3d5308b1e9f4c682ef510fceefec78b626b49)

File: src/app/models/feedback.py (hunk 5 above)
Comment (RusticPotatoes) [reply src/app/models/feedback.py]: resolved with [e260e5f](https://github.com/Bot-detector/public-api/pull/25/commits/e260e5f534fa87252579a96a232b701f1f35906d)

File: mysql/docker-entrypoint-initdb.d/02_data.sql (hunk 3 above)
Comment (RusticPotatoes) [reply mysql/docker-entrypoint-initdb.d/02_data.sql]: resolved by
[remove unneeded select](https://github.com/Bot-detector/public-api/pull/25/commits/de1cdba08613b58f5163d666e6f03b5f5d659933)

File: mysql/docker-entrypoint-initdb.d/02_data.sql (hunk 2 above)
Comment (RusticPotatoes) [reply mysql/docker-entrypoint-initdb.d/02_data.sql]: moved 
[move create table to 01_tables.sql](https://github.com/Bot-detector/public-api/pull/25/commits/75f4cbf2abc95b58bf0ac9641be15083a3cfdb1c)

File: src/app/views/input/feedback.py (hunk 6 above)
Comment (RusticPotatoes) [reply src/app/views/input/feedback.py]: resolved [removing feedback out](https://github.com/Bot-detector/public-api/pull/25/commits/909b9bca5580c5d74cf9607a3e9571833f8f9e2a)

File: mysql/docker-entrypoint-initdb.d/02_data.sql (hunk 1 above)
Comment (RusticPotatoes) [reply mysql/docker-entrypoint-initdb.d/02_data.sql]: uhhhhh so i did this - 

https://github.com/RusticPotatoes/public-api/blob/909b9bca5580c5d74cf9607a3e9571833f8f9e2a/mysql/docker-entrypoint-initdb.d/01_tables.sql#L17-L19

```mysql
    -- Add foreign keys for feedback_received and feedback_given
    feedback_received INT,
    feedback_given INT
```

https://github.com/RusticPotatoes/public-api/blob/909b9bca5580c5d74cf9607a3e9571833f8f9e2a/mysql/docker-entrypoint-initdb.d/01_tables.sql#L114C1-L119

```mysql
-- Add foreign key constraints to Players table
ALTER TABLE Players
ADD CONSTRAINT `FK_Feedback_Received` FOREIGN KEY (`feedback_received`) REFERENCES `PredictionFeedback` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT;

ALTER TABLE Players
ADD CONSTRAINT `FK_Feedback_Given` FOREIGN KEY (`feedback_given`) REFERENCES `PredictionFeedback` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT;
```

Comment (RusticPotatoes) [reply mysql/docker-entrypoint-initdb.d/02_data.sql]: resolved by

[remove foreign keys not needed for given and received](https://github.com/Bot-detector/public-api/pull/25/commits/ab9a14783294c2231895190c10eaa89a7d659b1a)
[ab9a147](https://github.com/Bot-detector/public-api/pull/25/commits/ab9a14783294c2231895190c10eaa89a7d659b1a)

and 

[remove PredictionFeedbackOut and foreign keys](https://github.com/Bot-detector/public-api/pull/25/commits/fa684a8c63e7611eacd3c8e1b5b4c4eb3f9dc405)

as tests are still passing after removal

File: mysql/docker-entrypoint-initdb.d/02_data.sql [hunk 7]
```diff
@@ -288,72 +118,133 @@ LIMIT
...
-	SET i = i + 1;
-	END WHILE;
-	END$$ 
+    DECLARE i INT;
+    SET i = 1;
+    WHILE i <= num DO SET @rand = FLOOR(1 + RAND() * 25);
+    SET @multiplier = FLOOR(51 + RAND() * 50);
+    INSERT INTO Predictions
+    SET
+        name = CONCAT('player', i),
```
Comment (extreme4all) [inline mysql/docker-entrypoint-initdb.d/02_data.sql]: good catch on this btw

File: src/api/v2/feedback.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    Raises:
+        HTTPException: Returns a 404 error with the message "Player not found" if no data is found for the user.
+
+    """
+    feedback = Feedback(session)
+    names = await asyncio.gather(*[to_jagex_name(n) for n in name])
+    data = await feedback.get_feedback(player_names=names)
+
+    return data
```
Comment (extreme4all) [inline src/api/v2/feedback.py]: we may not want to publicly expose this

File: src/api/v2/player.py [hunk 9]
```diff
@@ -13,7 +13,7 @@
 
 @router.get("/players/score", response_model=list[KCResponse])
 async def get_players_kc(
-    name: Annotated[list[str], Query(..., max_length=13)], session=Depends(get_session)
+    name: Annotated[list[str], Query(...)], session=Depends(get_session)
```
Comment (extreme4all) [inline src/api/v2/player.py]: any reason why max_length is removed here?

File: src/api/v2/player.py [hunk 10]
```diff
@@ -13,7 +13,7 @@
 
 @router.get("/players/score", response_model=list[KCResponse])
```
Comment (extreme4all) [inline src/api/v2/player.py]: can we rename to /players/report/score

File: src/app/models/feedback.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            query = query.join(
+                feedback_subject, feedback_subject.subject_id == player_name.id
+            )
+            query = query.join(
+                feedback_voter, feedback_voter.voter_id == player_name.id
+            )
+            query = query.where(
+                text("player_name.name IN :names").bindparams(
+                    names=player_names
+                )  # prevent sql injection
```
Comment (extreme4all) [inline src/app/models/feedback.py]: i think it is better to not "raw dog" sql
```
query.where(dbPlayer.name.in_(player_names))
```

File: src/app/models/feedback.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                    feedback_voter.proposed_label,
+                ]
+            )
+            query = query.select_from(player_name)
+            query = query.join(
+                feedback_subject, feedback_subject.subject_id == player_name.id
+            )
+            query = query.join(
+                feedback_voter, feedback_voter.voter_id == player_name.id
+            )
```
Comment (extreme4all) [inline src/app/models/feedback.py]: your joins are the wrong way around, which will result in adverse effect
```
select 
  pl.*, 
  feedback_subject.*, 
  feedback_voter.*
from Players pl
join Feedback feedback_subject on pl.id = feedback_subject.subject_id
join Feedback feedback_voter on pl.id = feedback_voter.voter_id
```
instead of joining the Players table twice (for voter & subject)

File: src/app/models/feedback.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            query = query.where(
+                text("player_name.name IN :names").bindparams(
+                    names=player_names
+                )  # prevent sql injection
+            )
```
Comment (extreme4all) [inline src/app/models/feedback.py]: this should be the query.

note for semantic reasons in the join we set the left of the "==" to the table we are joining to: "dbFeedback" because the default behaviour of a join is a left join.

```sql
feedback_voter: dbPlayer = aliased(dbPlayer, name="feedback_voter")
feedback_subject: dbPlayer = aliased(dbPlayer, name="feedback_subject")

query: Select = select(
	[
		func.count(func.distinct(feedback_subject.id)).label("count"),
		feedback_subject.possible_ban,
		feedback_subject.confirmed_ban,
		feedback_subject.confirmed_player,
		dbFeedback.vote,
	]
)
query = query.select_from(dbFeedback)
query = query.join(
	feedback_voter, dbFeedback.voter_id == feedback_voter.id
)
query = query.join(
	feedback_subject, dbFeedback.subject_id == feedback_subject.id
)
query = query.where(feedback_voter.name.in_(player_names))
query = query.group_by(
	feedback_subject.possible_ban,
	feedback_subject.confirmed_ban,
	feedback_subject.confirmed_player,
	dbFeedback.vote,
)
```

File: src/app/models/feedback.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+from src.core.database.models.feedback import DataModelPredictionFeedback as dbFeedback
+from src.core.database.models.player import Player as dbPlayer
+
+
+class Feedback:
+    def __init__(self, session: AsyncSession) -> None:
+        self.session = session
+        # self.logger = logging.getLogger("Feedback")
+
+    async def get_feedback(self, player_names: list[str]):
```
Comment (extreme4all) [inline src/app/models/feedback.py]: note that we may not want this query to be exposed to everyone "feedback_text" & "label" can be considered "sensitivee"

File: src/app/models/player.py [hunk 15]
```diff
@@ -32,29 +32,32 @@ async def get_kc(self, player_names: tuple[str]):
...
+                reporting_player, report_db.reportingID == reporting_player.id
             )
             query = query.join(
-                reported_player, dbReport.reportedID == reported_player.id
+                reported_player, report_db.reportedID == reported_player.id
             )
-            query = query.where(reporting_player.name.in_(player_names))
+            query = query.where(
+                text("reporting_player.name IN :names").bindparams(names=player_names)
+            )  # prevent sql injection
```
Comment (extreme4all) [inline src/app/models/player.py:55]: please revert this

File: src/app/models/player.py [hunk 16]
```diff
@@ -32,29 +32,32 @@ async def get_kc(self, player_names: tuple[str]):
...
             query = query.group_by(
                 reported_player.possible_ban,
                 reported_player.confirmed_ban,
                 reported_player.confirmed_player,
-                dbReport.manual_detect,
+                report_db.manual_detect,
             )
             result: Result = await self.session.execute(query)
             await self.session.commit()
```
Comment (extreme4all) [inline src/app/models/player.py:93]: Please revert the changes in this file

File: src/app/models/prediction.py [hunk 17]
```diff
@@ -38,14 +47,22 @@ async def get_prediction(self, user_name) -> list[dict]:
...
-        output = jsonable_encoder(output)
+        # verify prediction data exists
+        if all("Prediction" not in o for o in output):
+            output = [object_as_dict(o["prediction_db"]) for o in output]
+            output = jsonable_encoder(output)
+        # verify output is not empty list of none
+        if all(o is None for o in output):
+            return None
+
         return output
```
Comment (extreme4all) [inline src/app/models/prediction.py]: what are the reasons for these changes

File: src/app/models/prediction.py [hunk 18]
```diff
@@ -38,14 +47,22 @@ async def get_prediction(self, user_name) -> list[dict]:
...
         async with self.session:
-            query: Select = select(dbPrediction)
-            query = query.select_from(dbPrediction)
-            query = query.where(dbPrediction.name == user_name)
+            prediction_db: dbPrediction = aliased(dbPrediction, name="prediction_db")
+            query: Select = select(prediction_db)
+            query = query.select_from(prediction_db)
+            query = query.where(
+                text("prediction_db.name = :name").bindparams(name=user_name)
+            )  # prevent sql injection
```
Comment (extreme4all) [inline src/app/models/prediction.py]: please revert this change

File: src/app/views/response/feedback.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+    confidence: Optional[float]
+    feedback_text: Optional[str]
+    proposed_label: Optional[str]
+
+
+class FeedbackCount(BaseModel):
+    count: int
+    possible_ban: bool
+    confirmed_ban: bool
+    confirmed_player: bool
```
Comment (extreme4all) [inline src/app/views/response/feedback.py]: add vote to this

File: src/core/database/models/player.py [hunk 20]
```diff
@@ -19,3 +19,19 @@ class Player(Base):
...
+    # feedback_given = relationship(
+    #     "DataModelPredictionFeedback",
+    #     foreign_keys=[dbDataModelPredictionFeedback.voter_id],
+    #     back_populates="voter",
+    # )
+    # feedback_received = relationship(
+    #     "DataModelPredictionFeedback",
+    #     foreign_keys=[dbDataModelPredictionFeedback.subject_id],
+    #     back_populates="subject",
+    # )
```
Comment (extreme4all) [inline src/core/database/models/player.py]: remove comments please

File: src/api/v2/player.py (hunk 9 above)
Comment (RusticPotatoes) [reply src/api/v2/player.py]: length is not the length of the string, it's the length of the list. so it would limit how many players to send.  would have to integrate what I did in feedback if you want to add a str length limit

File: src/app/models/feedback.py (hunk 11 above)
Comment (RusticPotatoes) [reply src/app/models/feedback.py]: this is a strategy to prevent sql injection using the bind method

File: src/app/models/feedback.py (hunk 13 above)
Comment (RusticPotatoes) [reply src/app/models/feedback.py]: wait, there is a specific function for this, likely shouldn't expose the get_feedback as previously mentioned via the router in v2/feedback.py  https://github.com/Bot-detector/public-api/blob/4c1180aa78f47297df01280f7c21c719b9b79d35/src/app/models/feedback.py#L69-L120

File: src/app/models/player.py (hunk 16 above)
Comment (RusticPotatoes) [reply src/app/models/player.py:93]: I changed it to use the alias as it improves performance greatly is there another name structure you would like to use?
https://github.com/Bot-detector/public-api/blob/4c1180aa78f47297df01280f7c21c719b9b79d35/src/app/models/player.py#L35-L45

File: src/app/models/prediction.py (hunk 17 above)
Comment (RusticPotatoes) [reply src/app/models/prediction.py]: if the returned mappings is None we need logic to return None before getting to https://github.com/Bot-detector/public-api/blob/4c1180aa78f47297df01280f7c21c719b9b79d35/src/app/models/prediction.py#L62

otherwise you will get a list with None [None] so you need this to remediate.  

looking over we should just check the output value is not None otherwise return none so it never gets to object processing [None] list condition.

File: src/app/models/prediction.py (hunk 18 above)
Comment (RusticPotatoes) [reply src/app/models/prediction.py]: need to use bind params on any user inputted fields to avoid sql injection

File: src/app/models/feedback.py (hunk 11 above)
Comment (extreme4all) [reply src/app/models/feedback.py]: this also prevents sql injection and is a bit cleaner, using the orm
```
query.where(dbPlayer.name.in_(player_names))
```

File: src/api/v2/feedback.py (hunk 8 above)
Comment (RusticPotatoes) [reply src/api/v2/feedback.py]: so for the private api?

File: src/api/v2/player.py (hunk 9 above)
Comment (extreme4all) [reply src/api/v2/player.py]: oh interesting good catch! yeah we need to chech the lenght of each str, player names equal or longer than 13 char are not possible

Comment (RusticPotatoes) [reply src/api/v2/player.py]: will add this 
```python
    name: list[Annotated[str, Field(..., min_length=1, max_length=13)]] = Query(
        ...,
        min_length=1,
        description="Name of the player",
        example=["Player1", "Player2"],
    )
```

File: src/api/v2/player.py (hunk 10 above)
Comment (RusticPotatoes) [reply src/api/v2/player.py]: [str between 1-13 and at least 1 str, player/report/score](https://github.com/Bot-detector/public-api/pull/25/commits/1cf961c360e63de882247f3357a5f881cc2684ea)

File: src/api/v2/feedback.py (hunk 8 above)
Comment (extreme4all) [reply src/api/v2/feedback.py]: maybe in the future but probably not tbh

File: src/api/v2/player.py (hunk 9 above)
Comment (RusticPotatoes) [reply src/api/v2/player.py]: [str between 1-13 and at least 1 str, player/report/score](https://github.com/Bot-detector/public-api/pull/25/commits/1cf961c360e63de882247f3357a5f881cc2684ea)

File: src/app/models/feedback.py [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+                    reported_player.confirmed_ban,
+                    reported_player.confirmed_player,
+                ]
+            )
+            query = query.join(
+                reported_player, subquery.c.voter_id == reported_player.id
+            )
+            query = query.where(
+                text("reported_player.name IN :names").bindparams(names=player_names)
+            )  # prevent sql injection
```
Comment (extreme4all) [inline src/app/models/feedback.py]: this should be the query.

note for semantic reasons in the join we set the left of the "==" to the table we are joining to: "dbFeedback" because the default behaviour of a join is a left join.

```sql
feedback_voter: dbPlayer = aliased(dbPlayer, name="feedback_voter")
feedback_subject: dbPlayer = aliased(dbPlayer, name="feedback_subject")

query: Select = select(
	[
		func.count(func.distinct(feedback_subject.id)).label("count"),
		feedback_subject.possible_ban,
		feedback_subject.confirmed_ban,
		feedback_subject.confirmed_player,
		dbFeedback.vote,
	]
)
query = query.select_from(dbFeedback)
query = query.join(
	feedback_voter, dbFeedback.voter_id == feedback_voter.id
)
query = query.join(
	feedback_subject, dbFeedback.subject_id == feedback_subject.id
)
query = query.where(feedback_voter.name.in_(player_names))
query = query.group_by(
	feedback_subject.possible_ban,
	feedback_subject.confirmed_ban,
	feedback_subject.confirmed_player,
	dbFeedback.vote,
)
```

File: src/app/models/player.py (hunk 16 above)
Comment (extreme4all) [reply src/app/models/player.py:93]: Interesting, i don't immediatly understand how it would speed up to query speed.
what is the generated sql before and after the change?

File: src/app/models/prediction.py (hunk 18 above)
Comment (extreme4all) [reply src/app/models/prediction.py]: using the orm avoids sql injection, bind param are only relevant if you are writing raw sql like you do with text("")

File: src/core/database/models/player.py (hunk 20 above)
Comment (RusticPotatoes) [reply src/core/database/models/player.py]: [cleanup](https://github.com/Bot-detector/public-api/pull/25/commits/50b2d91281212cac50307dbea4ee5ec57b31840b)

File: src/app/models/feedback.py (hunk 11 above)
Comment (RusticPotatoes) [reply src/app/models/feedback.py]: [refactor player_name alias to player_db, remove bindparams](https://github.com/Bot-detector/public-api/pull/25/commits/91bdf391df482e4511ec664118865f178383a9cc)

File: src/app/models/feedback.py (hunk 21 above)
Comment (RusticPotatoes) [reply src/app/models/feedback.py]: this should be resolved, however this is the wrong function (get_feedback vs get feedback score)
get_feedback resolution:
https://github.com/Bot-detector/public-api/blob/91bdf391df482e4511ec664118865f178383a9cc/src/app/models/feedback.py#L38-L41

this is for get_feedback_score:
https://github.com/Bot-detector/public-api/blob/91bdf391df482e4511ec664118865f178383a9cc/src/app/models/feedback.py#L89-L92

note the subquery is defined here:
https://github.com/Bot-detector/public-api/blob/91bdf391df482e4511ec664118865f178383a9cc/src/app/models/feedback.py#L70-L79

Comment (RusticPotatoes) [reply src/app/models/feedback.py]: this is the new join for feedback score fixed
[name consistancy and correct join](https://github.com/Bot-detector/public-api/pull/25/commits/8fd4f95f35d41ded9129428567b3b6a34dd49dc4)

File: src/app/models/prediction.py (hunk 18 above)
Comment (RusticPotatoes) [reply src/app/models/prediction.py]: resolve with this commit - 
[prediction user_name to name, remove bind, revert list match, fix test cases](https://github.com/Bot-detector/public-api/pull/25/commits/d6ea27becdde54111f6d60eab2770e3c49e73839)

File: src/app/models/prediction.py (hunk 17 above)
Comment (RusticPotatoes) [reply src/app/models/prediction.py]: resolved with this commit - 
[prediction user_name to name, remove bind, revert list match, fix tes…](https://github.com/Bot-detector/public-api/pull/25/commits/d6ea27becdde54111f6d60eab2770e3c49e73839)

File: src/app/models/player.py (hunk 16 above)
Comment (RusticPotatoes) [reply src/app/models/player.py:93]: yea. they are the exact same in cost.  i may have been conflating the performance with using aliases with the performance i gained with feedback score subquery command.

```mysql
SELECT count(distinct(reported_player.id)) AS count, reported_player.possible_ban, reported_player.confirmed_ban, reported_player.confirmed_player, report_db.manual_detect 
FROM `Players` AS player_db, `Reports` AS report_db INNER JOIN `Players` AS reporting_player ON report_db.`reportingID` = reporting_player.id INNER JOIN `Players` AS reported_player ON report_db.`reportedID` = reported_player.id 
WHERE player_db.name IN ('player1', 'player2', 'player3') GROUP BY reported_player.possible_ban, reported_player.confirmed_ban, reported_player.confirmed_player, report_db.manual_detect

-- 50ms through 82ms

SELECT count(distinct(reported_player.id)) AS count, reported_player.possible_ban, reported_player.confirmed_ban, reported_player.confirmed_player, `Reports`.manual_detect 
FROM `Players` AS player_db, `Reports` INNER JOIN `Players` AS reporting_player ON `Reports`.`reportingID` = reporting_player.id INNER JOIN `Players` AS reported_player ON `Reports`.`reportedID` = reported_player.id 
WHERE player_db.name IN ('player1', 'player2', 'player3') GROUP BY reported_player.possible_ban, reported_player.confirmed_ban, reported_player.confirmed_player, `Reports`.manual_detect

-- 50ms through 82ms
```

File: src/api/v2/feedback.py (hunk 8 above)
Comment (RusticPotatoes) [reply src/api/v2/feedback.py]: where would we stash it?

File: src/app/models/player.py (hunk 16 above)
Comment (RusticPotatoes) [reply src/app/models/player.py:93]: [cleanup aliases](https://github.com/Bot-detector/public-api/pull/25/commits/c29eb904c219cbc01906f1cc27c6174b3378cc50)

File: src/app/views/response/feedback.py (hunk 19 above)
Comment (RusticPotatoes) [reply src/app/views/response/feedback.py]: TODO

File: src/app/models/feedback.py (hunk 14 above)
Comment (RusticPotatoes) [reply src/app/models/feedback.py]: do we just take away the route and leave this here for now?

File: src/app/views/response/feedback.py (hunk 19 above)
Comment (RusticPotatoes) [reply src/app/views/response/feedback.py]: > add vote to this

count is the summation of votes see
https://github.com/Bot-detector/public-api/blob/4c1180aa78f47297df01280f7c21c719b9b79d35/src/app/models/feedback.py#L79

File: src/app/models/feedback.py (hunk 14 above)
Comment (extreme4all) [reply src/app/models/feedback.py]: for now that is good yeah

File: mysql/docker-entrypoint-initdb.d/02_data.sql [hunk 22]
```diff
@@ -18,196 +17,41 @@ INSERT INTO
...
-                UNION
-                SELECT 7
-                UNION
-                SELECT 8
-                UNION
-                SELECT
-                    9
-            ) AS b
-    ) AS numbers;
-
```
Comment (extreme4all) [inline mysql/docker-entrypoint-initdb.d/02_data.sql:210]: seems to be messing with the data

File: src/app/views/input/feedback.py [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+            "zalcano_bot",
+            "woodcutting_bot",
+            "thieving_bot",
+            "soul_wars_bot",
+            "cooking_bot",
+            "vorkath_bot",
+            "barrows_bot",
+            "herblore_bot",
+            "unknown_bot",
+        ],
```
Comment (extreme4all) [inline src/app/views/input/feedback.py:64]: this is nice but having to maintain these lists in all places suck...

**Review by extreme4all — APPROVED** [review]
looks good
